coderd/notifications/notifier.go

```diff
@@ -0,0 +1,234 @@
+	cfg   codersdk.NotificationsConfig
+	ctx   context.Context
+	log   slog.Logger
+	store Store
+
+	tick     *time.Ticker
+	stopOnce sync.Once
+	quit     chan any
+	done     chan any
+
+	handlers *HandlerRegistry
+}
+
+func newNotifier(ctx context.Context, cfg codersdk.NotificationsConfig, id int, log slog.Logger, db Store, hr *HandlerRegistry) *notifier {
+	return &notifier{
+		id:       id,
+		ctx:      ctx,
+		cfg:      cfg,
+		log:      log.Named("notifier").With(slog.F("id", id)),
+		quit:     make(chan any),
+		done:     make(chan any),
+		tick:     time.NewTicker(cfg.FetchInterval.Value()),
+		store:    db,
+		handlers: hr,
+	}
+}
+
+// run is the main loop of the notifier.
+func (n *notifier) run(ctx context.Context, success chan<- dispatchResult, failure chan<- dispatchResult) error {
+	defer func() {
+		close(n.done)
+		n.log.Info(context.Background(), "gracefully stopped")
+	}()
+
+	// TODO: idea from Cian: instead of querying the database on a short interval, we could wait for pubsub notifications.
+	//		 if 100 notifications are enqueued, we shouldn't activate this routine for each one; so how to debounce these?
+	//		 PLUS we should also have an interval (but a longer one, maybe 1m) to account for retries (those will not get
+	//		 triggered by a code path, but rather by a timeout expiring which makes the message retryable)
+	for {
+		select {
+		case <-ctx.Done():
+			return xerrors.Errorf("notifier %d context canceled: %w", n.id, ctx.Err())
+		case <-n.quit:
+			return nil
+		default:
+		}
+
+		// Call process() immediately (i.e. don't wait an initial tick).
+		err := n.process(ctx, success, failure)
+		if err != nil {
+			n.log.Error(ctx, "failed to process messages", slog.Error(err))
+		}
+
+		// Shortcut to bail out quickly if stop() has been called or the context canceled.
+		select {
+		case <-ctx.Done():
+			return xerrors.Errorf("notifier %d context canceled: %w", n.id, ctx.Err())
+		case <-n.quit:
+			return nil
+		case <-n.tick.C:
+			// sleep until next invocation
+		}
+	}
+}
+
+// process is responsible for coordinating the retrieval, processing, and delivery of messages.
+// Messages are dispatched concurrently, but they may block when success/failure channels are full.
+//
+// NOTE: it is _possible_ that these goroutines could block for long enough to exceed CODER_NOTIFICATIONS_DISPATCH_TIMEOUT,
+// resulting in a failed attempt for each notification when their contexts are canceled; this is not possible with the
+// default configurations but could be brought about by an operator tuning things incorrectly.
+func (n *notifier) process(ctx context.Context, success chan<- dispatchResult, failure chan<- dispatchResult) error {
+	n.log.Debug(ctx, "attempting to dequeue messages")
+
+	msgs, err := n.fetch(ctx)
+	if err != nil {
+		return xerrors.Errorf("fetch messages: %w", err)
+	}
+
+	n.log.Debug(ctx, "dequeued messages", slog.F("count", len(msgs)))
+	if len(msgs) == 0 {
+		return nil
+	}
+
+	var eg errgroup.Group
```

Comment:
So, we go through a lot of effort piping these errors up from the delivery function, then collecting them in an error group, and the end result is that the call of this function just logs the error.

Instead, we should just log the errors down in the delivery functions and not return errors at all.

Reply:
I'm really just using this here as a slightly more ergonomic `sync.WaitGroup`. I can change it to that if you feel strongly about it. Failures are already logged for each dispatch.